src/main/java/com/metamx/emitter/core/HttpPostEmitter.java

```diff
@@ -166,315 +191,401 @@ public void emit(Event event)
-     */
-    private byte[] serializeBatch(List<byte[]> messages)
+    private void emitLargeEvents()
     {
-      final ByteArrayOutputStream baos = new ByteArrayOutputStream();
-      try {
-        serializeTo(messages, baos);
-        return baos.toByteArray();
+      if (largeEventsToEmit.isEmpty()) {
+        return;
       }
-      catch (IOException e) {
-        // There's no reason to have IOException in the signature of this method, since BAOS won't throw them.
-        throw Throwables.propagate(e);
+      // Don't try to emit large events until exhaustion, to avoid starvation of "normal" batches, if large event
+      // posting rate is too high, though it should never happen in practice.
+      largeEventsToEmit.add(LARGE_EVENTS_STOP);
+      for (byte[] largeEvent; (largeEvent = largeEventsToEmit.poll()) != LARGE_EVENTS_STOP; ) {
+        emitLargeEvent(largeEvent);
       }
     }
 
-    /**
-     * Serialize and compress (gzip) messages into a batch.
-     *
-     * @param messages list of JSON objects, one per message
-     * @return serialized and compressed JSON array
-     */
-    private byte[] serializeAndCompressBatch(List<byte[]> messages) throws IOException
+    private void emitLargeEvent(byte[] eventBytes)
     {
-      final ByteArrayOutputStream baos = new ByteArrayOutputStream();
-      try (
-          final GZIPOutputStream gzos = new GZIPOutputStream(baos)
-      ) {
-        serializeTo(messages, gzos);
+      byte[] buffer = acquireBuffer();
+      int bufferOffset = batchingStrategy.writeBatchStart(buffer);
+      System.arraycopy(eventBytes, 0, buffer, bufferOffset, eventBytes.length);
+      bufferOffset += eventBytes.length;
+      bufferOffset = batchingStrategy.writeBatchEnd(buffer, bufferOffset);
+      if (sendWithRetries(buffer, bufferOffset, 1)) {
+        buffersToReuse.add(buffer);
+      } else {
+        failedBuffers.add(new FailedBuffer(buffer, bufferOffset, 1));
       }
-      return baos.toByteArray();
     }
 
-    /**
-     * Serialize messages into a batch
-     *
-     * @param messages list of JSON objects, one per message
-     * @param os output stream with serialized batch
-     * @throws IOException if an I/O error occurs.
-     */
-    private void serializeTo(List<byte[]> messages, OutputStream os) throws IOException
+    private void tryEmitOneFailedBuffer()
     {
-      boolean first = true;
-      os.write(config.getBatchingStrategy().batchStart());
-      for (final byte[] message : messages) {
-        if (first) {
-          first = false;
-        } else {
-          os.write(config.getBatchingStrategy().messageSeparator());
+      FailedBuffer failedBuffer = failedBuffers.peek();
+      if (failedBuffer != null) {
+        if (sendWithRetries(failedBuffer.buffer, failedBuffer.length, failedBuffer.eventCount)) {
+          // Remove from the queue of failed buffer.
+          failedBuffers.poll();
+          buffersToReuse.add(failedBuffer.buffer);
```

Comment:
In a situation when we were not able to emit events for a while we don't have a way to discard buffers that were used to accumulate events during this period. For instance it may result in having 100 buffers in rotation even if we need just 2. 
One of the options to address it might be not add failed buffer back to buffersToReuse when it was finally sent.